checkmedia.c

```diff
@@ -1,598 +1,205 @@
 
+void help(void);
+int progress(unsigned percent);
 
 struct {
-  struct {
-    digest_ctx_t ctx;
-    digest_ctx_t full_ctx;
-    digest_t type;                              /* digest type */
-    char *name;                                 /* digest name */
-    int size;                                   /* digest size */
-    unsigned got_old:1;                         /* got digest stored in iso */
-    unsigned got_current:1;                     /* calculated current digest */
-    unsigned ok:1;                              /* digest matches */
-    unsigned char old[MAX_DIGEST_SIZE];         /* digest stored in iso */
-    unsigned char current[MAX_DIGEST_SIZE];     /* digest of iso ex special area */
-    unsigned char full[MAX_DIGEST_SIZE];        /* full digest of iso */
-  } digest;
-
-  unsigned err:1;		/* some error */
-  unsigned err_ofs;		/* read error pos */
-  unsigned size;		/* in kb */
-  unsigned hybrid_size;		/* hybrid partition, in kb */
-  unsigned media_nr;		/* media number */
-
-  char *media_type;		/* media type */
-  char vol_id[33];		/* volume id */
-  char app_id[81];		/* application id */
-  char app_data[0x201];		/* app specific data*/
-  unsigned pad;			/* pad size in kb */
-  unsigned pad_is_skip:1;	/* skip last sectors */
-} iso;
+  unsigned verbose:1;
+  unsigned help:1;
+  unsigned version:1;
+  char *file_name;
+} opt;
+
+struct option options[] = {
+  { "help", 0, NULL, 'h' },
+  { "verbose", 0, NULL, 'v' },
+  { "version", 0, NULL, 1 },
+  { }
+};
 
 
 int main(int argc, char **argv)
 {
   int i;
-  unsigned opt_verbose = 0;
-  unsigned opt_help = 0;
-  unsigned opt_version = 0;
-
-  if(argc > 1) {
-    if(!strcmp(argv[1], "-v") || !strcmp(argv[1], "--verbose")) {
-      opt_verbose = 1;
-      argc--;
-      argv++;
-    }
-    else if(!strcmp(argv[1], "-h") || !strcmp(argv[1], "--help")) {
-      opt_help = 1;
-    }
-    else if(!strcmp(argv[1], "--version")) {
-      opt_version = 1;
-    }
-  }
+  mediacheck_t *media;
 
-  if(opt_help || argc != 2) {
-    printf(
-      "usage: checkmedia [options] iso\n"
-      "Check SUSE installation media.\n"
-      "Options:\n"
-      "  --verbose\tshow some more info\n"
-      "  --version\tdisplay version number\n"
-      "  --help\tshort help text\n"
-    );
+  opterr = 0;
 
-    return 1;
+  while((i = getopt_long(argc, argv, "hv", options, NULL)) != -1) {
+    switch(i) {
+      case 1:
+        opt.version = 1;
+        break;
+
+      case 'v':
+        opt.verbose = 1;
+        break;
+
+      default:
+        help();
+        return i == 'h' ? 0 : 1;
+    }
   }
 
-  if(opt_version) {
+  if(opt.version) {
     printf(VERSION "\n");
-
     return 0;
   }
 
-  get_info(argv[1], opt_verbose);
-
-  if(iso.err) {
-    printf("%s: not an iso\n", argv[1]);
-
-    return 1;
+  if(argc == optind + 1) {
+    opt.file_name = argv[optind];
   }
-
-  if(!iso.digest.type) {
-    printf("%s: no digest found\n", argv[1]);
-
+  else {
+    help();
```

Comment:
maybe write before help something like "Mandatory parameter missing" ? and above with non -h variant print something like "Unknown parameter"? I think it is more user friendly that just printing help ( and usually helps me figure out what is wrong. Also this error explanation should go to STDERR. As help is print to stdout I think it should be important to print to stderr in case some just send STDOUT to STDERR and redirect it from console.

Reply:
ok, added a line